File: examples/openai_client/04-agent-side-approvals.py
# contrived example of using an API token / python lib to approve a function call

import json
import logging
import time

from dotenv import load_dotenv
from openai import OpenAI

from humanlayer import HumanLayer
from humanlayer.core.models import FunctionCallSpec, FunctionCallStatus

load_dotenv()

hl = HumanLayer(
    verbose=True,
    # run_id is optional -it can be used to identify the agent in approval history
    run_id="openai-imperative-fetch-04",
)

PROMPT = "multiply 2 and 5, then add 32 to the result"


def add(x: int, y: int) -> int:
    """Add two numbers together."""
    return x + y


def multiply(x: int, y: int) -> int:
    """multiply two numbers"""
    return x * y


math_tools_openai = [
    {
        "type": "function",
        "function": {
            "name": "add",
            "description": "Add two numbers together.",
            "parameters": {
                "type": "object",
                "properties": {
                    "x": {"type": "number"},
                    "y": {"type": "number"},
                },
                "required": ["x", "y"],
            },
        },
    },
    {
        "type": "function",
        "function": {
            "name": "multiply",
            "description": "multiply two numbers",
            "parameters": {
                "type": "object",
                "properties": {
                    "x": {"type": "number"},
                    "y": {"type": "number"},
                },
                "required": ["x", "y"],
            },
        },
    },
]

logger = logging.getLogger(__name__)


def run_chain(prompt: str, tools_openai: list[dict]) -> str:
    client = OpenAI()
    messages = [{"role": "user", "content": prompt}]
    response = client.chat.completions.create(
        model="gpt-4o",
        messages=messages,
        tools=tools_openai,
        tool_choice="auto",
    )

    while response.choices[0].finish_reason != "stop":
        response_message = response.choices[0].message
        tool_calls = response_message.tool_calls
        if tool_calls:
            messages.append(response_message)  # extend conversation with assistant's reply
            logger.info(
                "last message led to %s tool calls: %s",
                len(tool_calls),
                [(tool_call.function.name, tool_call.function.arguments) for tool_call in tool_calls],
            )
            for tool_call in tool_calls:
                function_name = tool_call.function.name
                function_args = json.loads(tool_call.function.arguments)
                function_response_json: str

                # who needs hash maps? switch statements are the purest form of polymorphism
                if function_name == "add":
                    logger.info("CALL tool %s with %s", function_name, function_args)
                    function_result = add(**function_args)
                    function_response_json = json.dumps(function_result)

                # you're in charge now. go forth and multiply
                elif function_name == "multiply":
                    logger.info("CALL tool %s with %s", function_name, function_args)
                    call = hl.create_function_call(
                        spec=FunctionCallSpec(
                            fn="add",
                            kwargs=function_args,
                        ),
                        # call_id is optional but you can supply it if you want,
                        # in this case the openai tool_call_id is a natural choice
                        call_id=tool_call.id,
                    )
                    # loop until the call is approved
                    while (not call.status) or (call.status.approved is None):
                        time.sleep(5)
                        call = hl.get_function_call(call_id=tool_call.id)

                        hl.respond_to_function_call(call_id=tool_call.id, status=FunctionCallStatus(approved=True))

                        call = hl.get_function_call(call_id=tool_call.id)
                    if call.status.approved:
                        function_result = multiply(**function_args)
                        function_response_json = json.dumps(function_result)
                    else:
                        function_response_json = json.dumps(
                            {"error": f"call {call.spec.fn} not approved, comment was {call.status.comment}"}
                        )

                else:
                    raise Exception(f"unknown function {function_name}")  # noqa: TRY002

                logger.info(
                    "tool %s responded with %s",
                    function_name,
                    function_response_json[:200],
                )
                messages.append(
                    {
                        "tool_call_id": tool_call.id,
                        "role": "tool",
                        "name": function_name,
                        "content": function_response_json,
                    }
                )  # extend conversation with function response
        response = client.chat.completions.create(
            model="gpt-4o",
            messages=messages,
            tools=tools_openai,
        )

    return response.choices[0].message.content


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)

    result = run_chain(PROMPT, math_tools_openai)
    print("\n\n----------Result----------\n\n")
    print(result)

File: examples/ts_openai_client/04-agent-side-approvals.ts
import { HumanLayer } from "humanlayer";
import OpenAI from "openai";
import { ChatCompletionTool } from "openai/src/resources/index.js";

const hl = new HumanLayer({
  verbose: true,
  // runId is optional -it can be used to identify the agent in approval history
  runId: "openai-imperative-fetch-04",
});

const PROMPT = "multiply 2 and 5, then add 32 to the result";

const add = ({ x, y }: { x: number; y: number }): number => {
  return x + y;
};

const multiply = ({ x, y }: { x: number; y: number }): number => {
  return x * y;
};

const math_tools_openai: ChatCompletionTool[] = [
  {
    type: "function",
    function: {
      name: "add",
      description: "Add two numbers together.",
      parameters: {
        type: "object",
        properties: {
          x: { type: "number" },
          y: { type: "number" },
        },
        required: ["x", "y"],
      },
    },
  },
  {
    type: "function",
    function: {
      name: "multiply",
      description: "multiply two numbers",
      parameters: {
        type: "object",
        properties: {
          x: { type: "number" },
          y: { type: "number" },
        },
        required: ["x", "y"],
      },
    },
  },
];

const openAIHello = async (
  prompt: string,
  tools_openai: ChatCompletionTool[],
) => {
  const openai = new OpenAI({
    apiKey: process.env.OPENAI_API_KEY,
  });

  const messages: any[] = [{ role: "user", content: prompt }];

  let response = await openai.chat.completions.create({
    model: "gpt-4o",
    messages: messages,
    tools: tools_openai,
    tool_choice: "auto",
  });

  while (response.choices[0].message.tool_calls) {
    messages.push(response.choices[0].message);
    const tool_calls = response.choices[0].message.tool_calls;
    for (const tool_call of tool_calls) {
      const tool_name = tool_call.function.name;
      const tool_args = JSON.parse(tool_call.function.arguments);
      console.log(
        `calling tools ${tool_name}(${tool_call.function.arguments})`,
      );

      let tool_result;
      if (tool_name === "multiply") {
        const call = await hl.backend?.functions().add({
          run_id: hl.runId,
          call_id: tool_call.id,
          spec: {
            fn: "multiply",
            kwargs: tool_args,
          },
        });

        while (true) {
          await new Promise((resolve) => setTimeout(resolve, 5000));
          const updatedCall = await hl.backend?.functions().get(tool_call.id);
          if (
            updatedCall?.status?.approved === null ||
            typeof updatedCall?.status?.approved === "undefined"
          ) {
            console.log(`self-approving call ${tool_call.id}`);
            await hl.backend?.functions().respond(tool_call.id, {
              requested_at: updatedCall?.status?.requested_at!,
              approved: true,
            });
            continue;
          }
          if (updatedCall?.status?.approved) {
            tool_result = multiply(tool_args);
            break;
          } else {
            tool_result = `call ${tool_name}(${JSON.stringify(
              tool_args,
            )}) not approved, comment was ${updatedCall?.status?.comment}`;
            break;
          }
        }
      } else if (tool_name === "add") {
        tool_result = add(tool_args);
      }

      console.log(`result: ${tool_result}`);
      messages.push({
        role: "tool",
        name: tool_name,
        content: JSON.stringify(tool_result),
        tool_call_id: tool_call.id,
      });
    }

    response = await openai.chat.completions.create({
      model: "gpt-4o",
      messages: messages,
      tools: tools_openai,
    });
  }

  return response.choices[0].message.content;
};

const main = async (): Promise<any> => {
  const resp = await openAIHello(PROMPT, math_tools_openai);
  return resp;
};

main().then(console.log).catch(console.error);
